File: backend/src/controllers/incident.controller.ts
import type { Request, Response } from 'express';
import { Incident } from '../models/Incident';
import { Evidence } from '../models/Evidence';
import { AuditRecord } from '../models/AuditRecord';
import { sendSuccess, sendError, ApiError } from '../utils/apiResponse';

export const createIncident = async (req: Request, res: Response) => {
  try {
    if (!req.user) {
      throw new ApiError(401, 'Authentication required.');
    }

    const { title, description, incidentType, severity, requiredSkills, location, evidence } = req.body;

    if (!evidence || !Array.isArray(evidence) || evidence.length === 0) {
      return sendError(res, 422, 'At least one evidence file is required.', [{ field: 'evidence', message: 'Evidence is mandatory.' }]);
    }

    const incident = await Incident.create({
      reporterId: req.user.id,
      incidentType,
      title,
      description,
      location: {
        type: 'Point',
        coordinates: [location.longitude, location.latitude],
        address: location.address
      },
      severity: severity || 'Medium',
      requiredSkills: requiredSkills || [],
      evidenceIds: [],
      status: 'PENDING_EVIDENCE',
      verificationStatus: 'UNVERIFIED'
    });

    const evidenceDocs = await Promise.all(
      evidence.map((file: any) =>
        Evidence.create({
          incidentId: incident._id,
          uploadedBy: req.user!.id,
          fileName: file.fileName,
          filePath: file.filePath,
          mimeType: file.mimeType,
          fileSize: file.fileSize,
          validationStatus: 'PENDING',
          metadata: { source: 'upload' }
        })
      )
    );

    incident.evidenceIds = evidenceDocs.map((doc) => doc._id);
    incident.status = 'EVIDENCE_VALIDATED';
    incident.verificationStatus = 'UNVERIFIED';
    await incident.save();

    await AuditRecord.create({
      actorId: req.user.id,
      actorRole: req.user.role,
      action: 'incident_creation',
      entityType: 'Incident',
      entityId: incident._id,
      metadata: {
        incidentType,
        title,
        status: incident.status,
        evidenceCount: evidenceDocs.length
      }
    });

    return sendSuccess(res, 201, 'Incident created successfully and moved into verification workflow.', {
      incident: {
        id: incident._id,
        title: incident.title,
        description: incident.description,
        type: incident.incidentType,
        severity: incident.severity,
        status: incident.status,
        verificationStatus: incident.verificationStatus,
        location: {
          latitude: incident.location.coordinates[1],
          longitude: incident.location.coordinates[0],
          address: incident.location.address
        },
        createdAt: incident.createdAt
      },
      evidence: evidenceDocs.map((doc) => ({
        id: doc._id,
        fileName: doc.fileName,
        mimeType: doc.mimeType,
        validationStatus: doc.validationStatus
      }))
    });
  } catch (error) {
    if (error instanceof ApiError) {
      return sendError(res, error.statusCode, error.message, error.details ?? []);
    }

    return sendError(res, 500, 'Unable to create incident.', [(error as Error).message]);
  }
};

export const listIncidents = async (_req: Request, res: Response) => {
  try {
    const incidents = await Incident.find().sort({ createdAt: -1 }).lean();

    return sendSuccess(res, 200, 'Incidents loaded.', {
      incidents: incidents.map((incident) => ({
        id: incident._id,
        title: incident.title,
        description: incident.description,
        type: incident.incidentType,
        severity: incident.severity,
        status: incident.status,
        verificationStatus: incident.verificationStatus,
        location: {
          latitude: incident.location.coordinates[1],
          longitude: incident.location.coordinates[0],
          address: incident.location.address
        },
        createdAt: incident.createdAt
      }))
    });
  } catch (error) {
    return sendError(res, 500, 'Unable to load incidents.', [(error as Error).message]);
  }
};

export const getIncidentById = async (req: Request, res: Response) => {
  try {
    const { incidentId } = req.params;
    const incident = await Incident.findById(incidentId).lean();

    if (!incident) {
      return sendError(res, 404, 'Incident not found.', [{ field: 'incidentId', message: 'No incident found with the provided id.' }]);
    }

    return sendSuccess(res, 200, 'Incident loaded.', {
      incident: {
        id: incident._id,
        title: incident.title,
        description: incident.description,
        type: incident.incidentType,
        severity: incident.severity,
        status: incident.status,
        verificationStatus: incident.verificationStatus,
        location: {
          latitude: incident.location.coordinates[1],
          longitude: incident.location.coordinates[0],
          address: incident.location.address
        },
        createdAt: incident.createdAt
      }
    });
  } catch (error) {
    return sendError(res, 500, 'Unable to load incident.', [(error as Error).message]);
  }
};

File: backend/src/main/java/com/raava/controller/IncidentController.java
package com.raava.controller;

import com.raava.dto.RaavaDtos;
import com.raava.exception.ApiException;
import com.raava.model.Incident;
import com.raava.model.User;
import com.raava.repository.IncidentRepository;
import com.raava.repository.UserRepository;
import com.raava.repository.AuditRecordRepository;
import com.raava.model.AuditRecord;
import com.raava.model.Evidence;
import com.raava.repository.EvidenceRepository;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.time.Instant;
import java.util.*;

@RestController
@RequestMapping("/api/incidents")
public class IncidentController {
    private final IncidentRepository incidentRepository;
    private final UserRepository userRepository;
    private final EvidenceRepository evidenceRepository;
    private final AuditRecordRepository auditRecordRepository;

    public IncidentController(IncidentRepository incidentRepository, UserRepository userRepository,
                             EvidenceRepository evidenceRepository, AuditRecordRepository auditRecordRepository) {
        this.incidentRepository = incidentRepository;
        this.userRepository = userRepository;
        this.evidenceRepository = evidenceRepository;
        this.auditRecordRepository = auditRecordRepository;
    }

    @GetMapping
    public ResponseEntity<Map<String, Object>> list() {
        List<Map<String, Object>> incidents = incidentRepository.findAll().stream().map(this::toSummary).toList();
        return ResponseEntity.ok(Map.of("success", true, "incidents", incidents));
    }

    @GetMapping("/{incidentId}")
    public ResponseEntity<Map<String, Object>> getById(@PathVariable String incidentId) {
        Incident incident = incidentRepository.findById(incidentId)
                .orElseThrow(() -> new ApiException(404, "Incident not found."));
        return ResponseEntity.ok(Map.of("success", true, "incident", toSummary(incident)));
    }

    @PostMapping
    public ResponseEntity<Map<String, Object>> create(@Valid @RequestBody RaavaDtos.CreateIncidentRequest request) {
        String userId = SecurityContextHolder.getContext().getAuthentication() == null
                ? null
                : (String) SecurityContextHolder.getContext().getAuthentication().getPrincipal();

        if (userId == null) {
            throw new ApiException(401, "Authentication required.");
        }

        User reporter = userRepository.findById(userId)
                .orElseThrow(() -> new ApiException(401, "User not found."));

        if (request.evidence() == null || request.evidence().isEmpty()) {
            throw new ApiException(422, "At least one evidence file is required.");
        }

        Incident incident = Incident.builder()
                .reporter(reporter)
                .incidentType(request.incidentType())
                .title(request.title())
                .description(request.description())
                .location(new Incident.GeoLocation("Point", new double[]{request.location().longitude(), request.location().latitude()}, request.location().address()))
                .severity("Low".equalsIgnoreCase(request.severity()) ? Incident.Severity.Low : "High".equalsIgnoreCase(request.severity()) ? Incident.Severity.High : Incident.Severity.Medium)
                .requiredSkills(request.requiredSkills() == null ? new ArrayList<>() : request.requiredSkills())
                .status(Incident.Status.EVIDENCE_VALIDATED)
                .verificationStatus(Incident.VerificationStatus.UNVERIFIED)
                .reportedAt(Instant.now())
                .createdAt(Instant.now())
                .updatedAt(Instant.now())
                .build();

        Incident savedIncident = incidentRepository.save(incident);

        List<Evidence> evidenceDocs = new ArrayList<>();
        for (RaavaDtos.EvidenceUpload evidenceUpload : request.evidence()) {
            Evidence evidence = Evidence.builder()
                    .incident(savedIncident)
                    .uploadedBy(reporter)
                    .fileName(evidenceUpload.fileName())
                    .filePath(evidenceUpload.filePath())
                    .mimeType(evidenceUpload.mimeType())
                    .fileSize(evidenceUpload.fileSize())
                    .validationStatus(Evidence.ValidationStatus.PENDING)
                    .metadata("{\"source\":\"upload\"}")
                    .createdAt(Instant.now())
                    .updatedAt(Instant.now())
                    .build();
            evidenceDocs.add(evidenceRepository.save(evidence));
        }

        savedIncident.setEvidence(evidenceDocs);
        incidentRepository.save(savedIncident);

        auditRecordRepository.save(AuditRecord.builder()
                .actor(reporter)
                .actorRole(reporter.getRole().name())
                .action("incident_creation")
                .entityType("Incident")
                .entityId(savedIncident.getId())
                .metadata(Map.of("title", savedIncident.getTitle(), "evidenceCount", evidenceDocs.size()))
                .createdAt(Instant.now())
                .updatedAt(Instant.now())
                .build());

        Map<String, Object> body = new HashMap<>();
        body.put("success", true);
        body.put("incident", toSummary(savedIncident));
        body.put("evidence", evidenceDocs.stream().map(this::toEvidenceSummary).toList());
        return ResponseEntity.status(201).body(body);
    }

    private Map<String, Object> toSummary(Incident incident) {
        Map<String, Object> summary = new HashMap<>();
        summary.put("id", incident.getId());
        summary.put("title", incident.getTitle());
        summary.put("description", incident.getDescription());
        summary.put("type", incident.getIncidentType());
        summary.put("severity", incident.getSeverity() == null ? null : incident.getSeverity().name());
        summary.put("status", incident.getStatus() == null ? null : incident.getStatus().name());
        summary.put("verificationStatus", incident.getVerificationStatus() == null ? null : incident.getVerificationStatus().name());
        summary.put("location", Map.of(
                "latitude", incident.getLocation() == null || incident.getLocation().getCoordinates() == null || incident.getLocation().getCoordinates().length < 2 ? null : incident.getLocation().getCoordinates()[1],
                "longitude", incident.getLocation() == null || incident.getLocation().getCoordinates() == null || incident.getLocation().getCoordinates().length < 2 ? null : incident.getLocation().getCoordinates()[0],
                "address", incident.getLocation() == null ? null : incident.getLocation().getAddress()
        ));
        summary.put("createdAt", incident.getCreatedAt());
        return summary;
    }

    private Map<String, Object> toEvidenceSummary(Evidence evidence) {
        Map<String, Object> summary = new HashMap<>();
        summary.put("id", evidence.getId());
        summary.put("fileName", evidence.getFileName());
        summary.put("mimeType", evidence.getMimeType());
        summary.put("validationStatus", evidence.getValidationStatus().name());
        return summary;
    }
}
